File: Feature Rich Onebox For Emails/src/services/elasticsearchService.ts
import { Client } from '@elastic/elasticsearch';
import { config } from '../config/config';
import { Email } from '../models/email';
import { logger } from '../utils/logger';

export class ElasticsearchService {
  private client: Client;
  private indexName: string;

  constructor() {
    this.indexName = config.elasticsearch.indexName;
    this.client = new Client({
      node: config.elasticsearch.node,
      auth: {
        username: config.elasticsearch.auth.username,
        password: config.elasticsearch.auth.password
      },
      tls: {
        rejectUnauthorized: false
      }
    });
  }

  // Initialize Elasticsearch service
  async initialize(): Promise<void> {
    try {
      // Check if Elasticsearch is available
      await this.client.ping();
      logger.info('Connected to Elasticsearch');

      // Check if index exists
      const indexExists = await this.client.indices.exists({ index: this.indexName });
      
      if (!indexExists) {
        await this.createIndex();
        logger.info(`Created index ${this.indexName}`);
      }
    } catch (error) {
      logger.error('Failed to connect to Elasticsearch:', error);
      throw error;
    }
  }

  // Create emails index with mappings
  private async createIndex(): Promise<void> {
    try {
      await this.client.indices.create({
        index: this.indexName,
        body: {
          settings: {
            number_of_shards: 1,
            number_of_replicas: 0,
            analysis: {
              analyzer: {
                email_analyzer: {
                  type: 'custom',
                  tokenizer: 'standard',
                  filter: ['lowercase', 'stop', 'snowball']
                }
              }
            }
          },
          mappings: {
            properties: {
              id: { type: 'keyword' },
              accountId: { type: 'keyword' },
              msgId: { type: 'keyword' },
              folder: { type: 'keyword' },
              uid: { type: 'long' },
              subject: { 
                type: 'text',
                analyzer: 'email_analyzer',
                fields: {
                  keyword: { type: 'keyword' }
                }
              },
              from: { 
                type: 'text',
                fields: {
                  keyword: { type: 'keyword' }
                }
              },
              to: { 
                type: 'text',
                fields: {
                  keyword: { type: 'keyword' }
                }
              },
              cc: { 
                type: 'text',
                fields: {
                  keyword: { type: 'keyword' }
                }
              },
              date: { type: 'date' },
              receivedDate: { type: 'date' },
              text: { 
                type: 'text',
                analyzer: 'email_analyzer' 
              },
              html: { type: 'text' },
              category: { type: 'keyword' },
              flags: { type: 'keyword' },
              attachments: {
                type: 'nested',
                properties: {
                  filename: { type: 'keyword' },
                  contentType: { type: 'keyword' },
                  size: { type: 'long' }
                }
              }
            }
          }
        }
      });
    } catch (error) {
      logger.error(`Error creating index ${this.indexName}:`, error);
      throw error;
    }
  }

  // Close Elasticsearch client
  async close(): Promise<void> {
    try {
      await this.client.close();
      logger.info('Elasticsearch client closed');
    } catch (error) {
      logger.error('Error closing Elasticsearch client:', error);
    }
  }

  // Index an email
  async indexEmail(email: Email): Promise<string> {
    try {
      // Check if email already exists
      const exists = await this.emailExists(email.id);
      
      if (exists) {
        logger.debug(`Email ${email.id} already exists, updating`);
        
        // Update existing email
        await this.client.update({
          index: this.indexName,
          id: email.id,
          body: {
            doc: email
          }
        });
        
        return email.id;
      }
      
      // Index new email
      const response = await this.client.index({
        index: this.indexName,
        id: email.id,
        body: email
      });
      
      logger.debug(`Indexed email ${email.id}`);
      return response._id;
    } catch (error) {
      logger.error(`Error indexing email ${email.id}:`, error);
      throw error;
    }
  }

  // Check if an email exists
  async emailExists(id: string): Promise<boolean> {
    try {
      const response = await this.client.exists({
        index: this.indexName,
        id
      });
      
      return response;
    } catch (error) {
      logger.error(`Error checking if email ${id} exists:`, error);
      return false;
    }
  }

  // Get an email by ID
  async getEmailById(id: string): Promise<Email | null> {
    try {
      const response = await this.client.get({
        index: this.indexName,
        id
      });
      
      if (!response || !response._source) {
        return null;
      }
      
      return response._source as Email;
    } catch (error) {
      if (error.meta?.statusCode === 404) {
        return null;
      }
      
      logger.error(`Error getting email ${id}:`, error);
      throw error;
    }
  }

  // Update email category
  async updateEmailCategory(id: string, category: string): Promise<void> {
    try {
      await this.client.update({
        index: this.indexName,
        id,
        body: {
          doc: {
            category
          }
        }
      });
      
      logger.debug(`Updated category for email ${id} to ${category}`);
    } catch (error) {
      logger.error(`Error updating category for email ${id}:`, error);
      throw error;
    }
  }

  // Search emails
  async searchEmails(options: any): Promise<any> {
    try {
      const { page = 1, limit = 50, query, sortBy = 'date', sortOrder = 'desc', filters = {} } = options;
      
      const from = (page - 1) * limit;
      
      // Build query
      const must: any[] = [];
      
      // Full-text search
      if (query) {
        must.push({
          multi_match: {
            query,
            fields: ['subject^3', 'from^2', 'to^2', 'text']
          }
        });
      }
      
      // Add filters
      if (filters.accountId) {
        must.push({ term: { accountId: filters.accountId } });
      }
      
      if (filters.folder) {
        must.push({ term: { folder: filters.folder } });
      }
      
      if (filters.category) {
        must.push({ term: { category: filters.category } });
      }
      
      if (filters.from) {
        must.push({ match: { from: filters.from } });
      }
      
      if (filters.to) {
        must.push({ match: { to: filters.to } });
      }
      
      if (filters.subject) {
        must.push({ match: { subject: filters.subject } });
      }
      
      // Date range
      if (filters.startDate || filters.endDate) {
        const dateRange: any = {};
        
        if (filters.startDate) {
          dateRange.gte = filters.startDate;
        }
        
        if (filters.endDate) {
          dateRange.lte = filters.endDate;
        }
        
        must.push({
          range: {
            date: dateRange
          }
        });
      }
      
      // Build sort
      const sort: any = {};
      sort[sortBy] = { order: sortOrder };
      
      // Perform search
      const body: any = {
        query: {
          bool: {
            must
          }
        },
        sort: [sort],
        from,
        size: limit
      };
      
      const response = await this.client.search({
        index: this.indexName,
        body
      });
      
      // Format results
      const total = response.hits.total.value;
      const emails = response.hits.hits.map((hit: any) => ({
        ...hit._source,
        score: hit._score
      }));
      
      return {
        total,
        page,
        limit,
        pages: Math.ceil(total / limit),
        emails
      };
    } catch (error) {
      logger.error('Error searching emails:', error);
      throw error;
    }
  }
}

File: Feature Rich Onebox For Emails/services/elasticsearch_service.py
import json
import logging
import os
from elasticsearch import Elasticsearch
from models import Email

logger = logging.getLogger(__name__)

class ElasticsearchService:
    """Service for handling Elasticsearch operations for email search and indexing"""
    
    def __init__(self):
        self.client = None
        self.index_name = 'emails'
        self.initialized = False
        
    def initialize(self):
        """Initialize Elasticsearch connection"""
        try:
            # Get Elasticsearch URL from environment or use default
            es_url = os.environ.get('ELASTICSEARCH_URL', 'http://localhost:9200')
            
            # Create Elasticsearch client
            self.client = Elasticsearch(es_url)
            
            # Check if connection is successful
            if self.client.ping():
                logger.info("Connected to Elasticsearch")
                
                # Create index if it doesn't exist
                self._create_index()
                self.initialized = True
                return True
            else:
                logger.error("Failed to connect to Elasticsearch")
                return False
        except Exception as e:
            logger.error(f"Elasticsearch initialization error: {str(e)}")
            return False
    
    def _create_index(self):
        """Create index with mappings if it doesn't exist"""
        if not self.client.indices.exists(index=self.index_name):
            mapping = {
                "mappings": {
                    "properties": {
                        "subject": {"type": "text"},
                        "body_text": {"type": "text"},
                        "sender": {"type": "text"},
                        "recipients": {"type": "text"},
                        "date": {"type": "date"},
                        "category": {"type": "keyword"},
                        "account_id": {"type": "keyword"},
                        "folder": {"type": "keyword"}
                    }
                }
            }
            
            self.client.indices.create(index=self.index_name, body=mapping)
            logger.info(f"Created Elasticsearch index '{self.index_name}'")
    
    def index_email(self, email):
        """Index an email in Elasticsearch"""
        if not self.initialized:
            self.initialize()
        
        if not self.initialized:
            logger.error("Cannot index email: Elasticsearch not initialized")
            return False
        
        try:
            doc = {
                'id': email.id,
                'account_id': email.account_id,
                'subject': email.subject,
                'body_text': email.body_text,
                'sender': email.sender,
                'recipients': email.recipients,
                'folder': email.folder,
                'date': email.date.isoformat() if email.date else None,
                'category': email.category
            }
            
            self.client.index(index=self.index_name, id=str(email.id), document=doc)
            logger.debug(f"Indexed email {email.id} in Elasticsearch")
            return True
        except Exception as e:
            logger.error(f"Error indexing email {email.id}: {str(e)}")
            return False
    
    def search_emails(self, options):
        """Search emails in Elasticsearch"""
        if not self.initialized:
            self.initialize()
        
        if not self.initialized:
            logger.error("Cannot search emails: Elasticsearch not initialized")
            return []
        
        try:
            query_string = options.get('query', '')
            filters = options.get('filters', {})
            
            # Build query
            query = {
                "bool": {
                    "must": []
                }
            }
            
            # Add text search if provided
            if query_string:
                query["bool"]["must"].append({
                    "multi_match": {
                        "query": query_string,
                        "fields": ["subject^2", "body_text", "sender", "recipients"]
                    }
                })
            
            # Add filters
            for field, value in filters.items():
                if value:
                    query["bool"]["must"].append({
                        "term": {field: value}
                    })
            
            # Execute search
            search_body = {
                "query": query,
                "sort": [{"date": {"order": "desc"}}],
                "size": 100
            }
            
            response = self.client.search(index=self.index_name, body=search_body)
            
            # Process results
            results = []
            for hit in response["hits"]["hits"]:
                results.append({
                    "id": hit["_source"]["id"],
                    "score": hit["_score"],
                    "subject": hit["_source"].get("subject", ""),
                    "sender": hit["_source"].get("sender", ""),
                    "date": hit["_source"].get("date")
                })
            
            return results
        except Exception as e:
            logger.error(f"Search error: {str(e)}")
            # If Elasticsearch fails, fall back to database query
            logger.info("Falling back to database query")
            return self._fallback_search(options)
    
    def _fallback_search(self, options):
        """Fall back to database query if Elasticsearch fails"""
        from app import db
        
        query_string = options.get('query', '')
        filters = options.get('filters', {})
        
        # Create base query
        query = db.session.query(Email)
        
        # Apply filters
        for field, value in filters.items():
            if value and hasattr(Email, field):
                query = query.filter(getattr(Email, field) == value)
        
        # Apply text search (simple LIKE query - not as effective as Elasticsearch)
        if query_string:
            search_term = f"%{query_string}%"
            query = query.filter(
                db.or_(
                    Email.subject.ilike(search_term),
                    Email.body_text.ilike(search_term),
                    Email.sender.ilike(search_term)
                )
            )
        
        # Get results
        emails = query.order_by(Email.date.desc()).limit(100).all()
        
        # Format results
        results = []
        for email in emails:
            results.append({
                "id": email.id,
                "score": 1.0,  # No real score in SQL query
                "subject": email.subject,
                "sender": email.sender,
                "date": email.date.isoformat() if email.date else None
            })
        
        return results
    
    def close(self):
        """Close Elasticsearch connection"""
        if self.client:
            logger.info("Closing Elasticsearch connection")
            self.client.close()